File: Week_03/kotlin/src/permutations/_47_PermutationsII.kt
package permutations

class _47_UsingSwap {
    fun permuteUnique(nums: IntArray): List<List<Int>> {
        val ret = mutableListOf<List<Int>>()
        val n = nums.size - 1
        fun dfs(start: Int) {
            if (start == n) {
                ret += nums.toList()
                return
            }
            val table = HashSet<Int>();
            /*loop@*/ for (i in start..n) {
                if (nums[i] in table) continue
                table += nums[i];
                // for (j in start until i) {
                //     if (nums[i] == nums[j]) continue@loop
                // }
                nums[i] = nums[start].also { nums[start] = nums[i] }
                dfs(start + 1);
                nums[i] = nums[start].also { nums[start] = nums[i] }
            }
        }
        dfs(0)
        return ret
    }
}

File: Week_03/cpp/src/permutations/_47_PermutationsII.cpp
#include <vector>
#include <iostream>

using namespace std;

class Solution
{
public:
    vector<vector<int>> ret;
    vector<vector<int>> permuteUnique(vector<int> &nums)
    {
        dfs(nums, 0);
        return ret;
    }
    void dfs(vector<int> &nums, int start)
    {
        if (start == nums.size() - 1)
        {
            ret.push_back(nums);
            return;
        }
        for (int i = start; i < nums.size(); i++)
        {
            bool duplicate = false;
            for (int j = start; j < i; j++)
            {
                if (nums[j] == nums[i])
                {
                    duplicate = true;
                    break;
                }
            }
            if (duplicate)
                continue;
            swap(nums[i], nums[start]);
            dfs(nums, start + 1);
            swap(nums[i], nums[start]);
        }
    }
};